src/env.h

```diff
@@ -719,8 +719,11 @@ class AsyncHooks : public MemoryRetainer {
   inline void no_force_checks();
   inline Environment* env();
 
+  // NB: This call does not take (co-)ownership of `execution_async_resource`.
+  // The lifetime of the `v8::Local<>` pointee must last until
+  // `pop_async_context()` or `clear_async_id_stack()` are called.
   inline void push_async_context(double async_id, double trigger_async_id,
-      v8::Local<v8::Object> execution_async_resource_);
+      v8::Local<v8::Object> execution_async_resource);
```

Comment:
If this doesn't take ownership of execution_async_resource, shouldn't we find a way to do this without assigning it to the elements of the native_execution_async_resources_ member?

Reply:
I don’t quite understand how the two are related, sorry. It’s not atypical for pointers or pointer-like objects to be used only for observational purposes.

Reply:
I was concerned about using non-owning pointers as member variables as they might turn into dangling pointers at some point, which are unsafe to dereference.

> It’s not atypical for pointers or pointer-like objects to be used only for observational purposes.

I'm completely fine with passing non-owning pointers through function calls as long as it is guaranteed that the call stacks get created and destroyed within the lifetime of the pointed variable but here we are dealing with sharing the pointer with other methods which might not share the same stack where the variable was created. Also, non-owning pointers are common as a member only for classes like std::string_view but is this class a view which is allocated and deallocated in between the lifetime of the locals?

"The lifetime of the `v8::Local<>` pointee must last until `pop_async_context()` or `clear_async_id_stack()` are called." sounds like it is merely a requirement for our code to work correctly. If it's something that happens for sure, shouldn't we add assertions everywhere we access the locals to make sure that they didn't get GCed?

Reply:
> I was concerned about using non-owning pointers as member variables as they might turn into dangling pointers at some point, which are unsafe to dereference.

Yeah, hence the warning in the comment here. I’m happy to take suggestions on the wording :)

> sounds like it is merely a requirement for our code to work correctly

Well yeah, that’s what it is. If this doesn’t happen, the code will break (i.e. potentially crash).

> If it's something that happens for sure, shouldn't we add assertions everywhere we access the locals to make sure that they didn't get GCed?

How would we check that? We could add weak `v8::Global`s references with callback that verify that the object is no longer on the stack here when it is GC’ed but that’s a) too invasive and too expensive to be done in non-debug mode and b) pretty complex.

Reply:
> Well yeah, that’s what it is. If this doesn’t happen, the code will break (i.e. potentially crash).

Ah, then the comment looks okay. Btw, I was trying to store a local and preserve it beyond the scope of the surrounding HandleScope in a binding to recreate this situation but it seems the address stored in the local gets changed for some reason? Do you know what I'm doing wrong here?

<details>

```cpp
#include <node.h>

void F(const v8::FunctionCallbackInfo<v8::Value>& args) {
  static v8::Local<v8::Value> local = args[0];
  args.GetReturnValue().Set(local);
}

void Initialize(v8::Local<v8::Object> exports) {
  NODE_SET_METHOD(exports, "f", F);
}

NODE_MODULE(ADDON, Initialize)
```

```js
const { f } = require('./build/Release/addon');
console.log(f({ 'hello': 'world' })); // this prints { hello: 'world' } as expected
console.log(f()); // this should either print { hello: 'world' } or crash if the object gets GCed but it prints the global object instead
```

```js
❯ node index.js
{ hello: 'world' }
<ref *1> Object [global] {
  global: [Circular *1],
  clearInterval: [Function: clearInterval],
  clearTimeout: [Function: clearTimeout],
  setInterval: [Function: setInterval],
  setTimeout: [Function: setTimeout] {
    [Symbol(nodejs.util.promisify.custom)]: [Getter]
  },
  queueMicrotask: [Function: queueMicrotask],
  performance: Performance {
    nodeTiming: PerformanceNodeTiming {
      name: 'node',
      entryType: 'node',
      startTime: 0,
      duration: 43.834795001894236,
      nodeStart: 2.549355000257492,
      v8Start: 3.2348700016736984,
      bootstrapComplete: 31.490434002131224,
      environment: 17.8702540025115,
      loopStart: -1,
      loopExit: -1,
      idleTime: 0
    },
    timeOrigin: 1640869141052.899
  },
  clearImmediate: [Function: clearImmediate],
  setImmediate: [Function: setImmediate] {
    [Symbol(nodejs.util.promisify.custom)]: [Getter]
  },
  structuredClone: [Function: structuredClone]
}
```

</details>

> We could add weak v8::Globals references with callback that verify that the object is no longer on the stack here when it is GC’ed but that’s a) too invasive and too expensive to be done in non-debug mode and b) pretty complex.

It's unfortunate that a simpler way doesn't exist for release builds. :/

Reply:
> Btw, I was trying to store a local and preserve it beyond the scope of the surrounding HandleScope in a binding to recreate this situation but it seems the address stored in the local gets changed for some reason? Do you know what I'm doing wrong here?

I don’t know, and it’s certainly unexpected to see the global object there, but I guess there’s no reason why it *wouldn’t* print the global object? As soon as you enter UB, anything can happen, including this … it probably just picked up something from another address at that point

Reply:
That makes sense but how is storing the local not UB in this PR? A local shouldn't be valid outside the scope of a HandleScope that was active during its creation, right?

Reply:
@RaisinTen Yeah, exactly. That’s why the comment is here telling people to make sure that the `Local`’s lifetime extends until it is removed from the stack.

Reply:
Okay, so since these functions can be called from the JS side using pushAsyncContext() and popAsyncContext(), accessing the locals between those JS calls would lead to UB because the HandleScopes inside these binding functions get popped from the stack in between calls and we are not doing any of that now. Is that right?

Reply:
The JS side never calls these functions with non-empty locals (otherwise this would be problematic, yes)